cleanrl/experiments/impl_matters/ppo_continuous_gae.py

```diff
@@ -0,0 +1,493 @@
+    ep_Returns = [] # Keeps track of the undiscounted return for each sampled episode
+
+    for step in range( args.episode_length):
+        global_step += 1
+        with torch.no_grad():
+            action, action_logp = pg.get_action( [obs])
+            action = action.numpy()[0]
+            action_logp = action_logp.numpy()[0]
+
+            v_obs =  vf.forward( [obs]).numpy()[0][0]
+
+        next_obs, rew, done, _ = env.step( action)
+
+        ep_obs.append( obs)
+        ep_acts.append( action)
+        ep_rews.append( rew)
+        ep_logps.append( action_logp)
+        ep_vals.append( v_obs)
+
+        obs = next_obs
+
+        if done:
+            # Updateing sampling stats
+            ep_count += 1
+            ep_lengths.append( step)
+            ep_Returns.append( np.sum( ep_rews))
+
+            ep_returns = discount_cumsum( ep_rews, args.gamma)
+
+            # Quick Hack GAE computation, which require np.array data
+            ep_rews = np.array( ep_rews)
+            ep_vals = np.array( ep_vals)
+            deltas = ep_rews[:-1] + args.gamma * ep_vals[1:] - ep_vals[:-1]
+            ep_vals = discount_cumsum(deltas, args.gamma * args.gae_lambda)
+
+            buffer.put_episode( zip( ep_obs, ep_acts, ep_logps, ep_returns, ep_vals))
+
+            # Cleanup the tmp holders for fresh episode data
+            ep_obs, ep_acts, ep_logps, ep_rews, ep_vals = [], [], [], [], []
+
+            obs = env.reset()
+            done = False
+
+        # Reached sampling limit, will cut the episode and use the value function
+        # to bootstrap, as per OpenAI SpinUp
+        if step == (args.episode_length - 1) and len(ep_obs) > 0:
+            # Updateing sampling stats, although they become "unexact"
+            ep_count += 1
+            ep_lengths.append( step)
+            ep_Returns.append( np.sum( ep_rews))
+
+            # NOTE: Last condition: if the episode finished right at the sampling
+            # limit, this will be skipped anyway
+            with torch.no_grad():
+                v_obs = vf.forward( [obs]).numpy()[0][0]
+
+            ep_rews[-1] = v_obs # Boostrapping reward
+
+            ep_returns = discount_cumsum( ep_rews, args.gamma)
+
+            # Quick Hack GAE computation, which require np.array data
+            ep_rews = np.array( ep_rews)
+            ep_vals = np.array( ep_vals)
+            deltas = ep_rews[:-1] + args.gamma * ep_vals[1:] - ep_vals[:-1]
+            ep_vals = discount_cumsum(deltas, args.gamma * args.gae_lambda)
+
+            buffer.put_episode( zip( ep_obs, ep_acts, ep_logps, ep_returns, ep_vals))
+
+    sampling_stats = {
+        "ep_count": ep_count,
+        "train_episode_length": np.mean( ep_lengths),
+        "train_episode_return": np.mean( ep_Returns)
+    }
+
+    return sampling_stats
+
+while global_step < args.total_timesteps:
+    # Will sample args.episode_length transitions, which we consider as the length of an epoch
+    sampling_stats = sample()
```

Comment:
This is related to the comment regarding the `global step`: the motivation was to separate the sampling process, which is just the data being collected, from the actual networks updates, as in the algorithm of the paper.

When I first read the ppo implementation that had a sampling method such as in the ppo_batch.py, it was actually harder to read for me and get a good overall idea of what was happening in the training. But again, it is purely subjective, so at this point, I do not mind using your method either.

Reply:
Oh I see. I think it’s a good intention, but the separation probably is not going to help us that much because we’re not using it to utilize the multi processing features. Given the amount of code that will be required to change on other files, I think it’s probably easier to just keep it simple.

Reply:
Roger.